### ISSUE TITLE
Failed to run memtest86+ 64bit EFI from FirmwareVolume

### ISSUE THREAD

**Issue by miczyg1** [memtest86plus/memtest86plus#377]
I am trying to integrate memtest86+ as a bootable option loaded from the UEFI FirmwareVolume in TianoCore UEFI payload. When the option is booted, I get a page fault with RIP pointing at this line: https://github.com/memtest86plus/memtest86plus/blob/main/boot/efisetup.c#L630

After analyzing the resulting memtest.efi binary, I noticed that the LD scripts do not contain the `.data` section and the EFI system table pointer is saved to a static variable that is in `.text` section with RO attributes (thus the page fault).

This is the debug output from UEFI Payload when the image is executed from FV:

```
InstallProtocolInterface: 5B1B31A1-9562-11D2-8E3F-00A0C969723B 45156040
Loading driver at 0x00000200000 EntryPoint=0x000002011E0 
InstallProtocolInterface: BC62157E-3E33-4FEC-9920-2D3B36D750DF 45174998
ProtectUefiImageCommon - 0x45156040
  - 0x0000000000200000 - 0x000000000006F000
SetUefiImageMemoryAttributes - 0x0000000000200000 - 0x0000000000001000 (0x0000000000004008)
SetUefiImageMemoryAttributes - 0x0000000000201000 - 0x0000000000024000 (0x0000000000020008)
SetUefiImageMemoryAttributes - 0x0000000000225000 - 0x000000000004A000 (0x0000000000004008)
```

BootServices marks the section as:
1. header as WB (writeback), XP (execute-protect)
2. text as WB, RO (read-only)
3. all the rest as WB, XP

The `sys_table` static variable lands at address 0x224BA0 (in my build) which is marked WB, RO. So memtest writes to read-only page. 

However, such situation doesn't happen when booted the memtest.iso target, but the flow of ProtectUefiImageCommon is also different there:

```
InstallProtocolInterface: 5B1B31A1-9562-11D2-8E3F-00A0C969723B 45156040
Loading driver at 0x00000200000 EntryPoint=0x000002011E0 
InstallProtocolInterface: BC62157E-3E33-4FEC-9920-2D3B36D750DF 45156E98
ProtectUefiImageCommon - 0x45156040
  - 0x0000000000200000 - 0x000000000006F000
```

So it seems to get this working, the memtest.efi would need to get a `.data` section outside `.text` section for clear separation.

**Comment by debrouxl** [comment]
Thanks for the report. Looks like something we'd need to fix.

Currently, as you have probably seen by now, the memtest86+ build system maintains a bunch of separate sections until memtest_shared, flattens memtest_shared to `.data` when using `objcopy -O binary` to produce memtest_shared.bin, then recreates a `.text` section from `.data` when producing memtest.efi. I didn't change that when I worked on #265 , for which I didn't send a pull request (yet).

**Comment by miczyg1** [comment]
@debrouxl my linker-fu isn't the best, but if you could give me some pointers how to fix it I would gladly patch things up and send a PR.

**Comment by debrouxl** [comment]
I'm not the best with linker scripts, either :)
By playing with objcopy arguments such as -j, I can certainly manage to produce three binaries: everything before the `.data` section, the `.data` section, and everything after it. But that still turns everything into `.data`, and reassembling these pieces later (how ?) is certainly the wrong approach. objcopy's pretty nice for having a proper memory image, though.

I tried to hack up a quick PoC for producing memtest.efi from memtest_shared:
```
diff --git a/boot/header.S b/boot/header.S
index 23e9a14..d331ce3 100644
--- a/boot/header.S
+++ b/boot/header.S
@@ -89,7 +89,7 @@ coff_header:
 #else
        .word   IMAGE_FILE_MACHINE_I386         # Machine (i386)
 #endif
-       .word   3                               # NumberOfSections
+       .word   4                               # NumberOfSections
        .long   0                               # TimeDateStamp
        .long   0                               # PointerToSymbolTable
        .long   0                               # NumberOfSymbols
@@ -193,6 +193,21 @@ section_table:
              | IMAGE_SCN_MEM_EXECUTE           \
              | IMAGE_SCN_CNT_CODE              # Characteristics (section flags)
 
+       .ascii  ".data"
+       .byte   0
+       .byte   0
+       .byte   0
+       .long   _virt_data_size                 # VirtualSize
+       .long   _virt_data_start                # VirtualAddress
+       .long   _file_data_size                 # SizeOfRawData
+       .long   _file_data_start                # PointerToRawData
+       .long   0                               # PointerToRelocations
+       .long   0                               # PointerToLineNumbers
+       .word   0                               # NumberOfRelocations
+       .word   0                               # NumberOfLineNumbers
+       .long   IMAGE_SCN_MEM_READ              \
+             | IMAGE_SCN_CNT_INITIALIZED_DATA  # Characteristics (section flags)
+
        .ascii  ".reloc"
        .byte   0
        .byte   0
diff --git a/build64/Makefile b/build64/Makefile
index 0445939..2f8f07c 100644
--- a/build64/Makefile
+++ b/build64/Makefile
@@ -172,9 +172,9 @@ memtest.bin: memtest_shared.bin boot/bootsect.o boot/setup.o ldscripts/memtest_b
        $(eval SIZES=$(shell size -B -d memtest_shared | grep memtest_shared))
        $(LD) --defsym=_bss_size=$(word 3,$(SIZES)) -T ldscripts/memtest_bin.lds boot/bootsect.o boot/setup.o -b binary memtest_shared.bin -o memtest.bin
 
-memtest.efi: memtest_shared.bin boot/header.o boot/setup.o ldscripts/memtest_efi.lds
+memtest.efi: memtest_shared boot/header.o boot/setup.o ldscripts/memtest_efi.lds
        $(eval SIZES=$(shell size -B -d memtest_shared | grep memtest_shared))
-       $(LD) --defsym=_bss_size=$(word 3,$(SIZES)) -T ldscripts/memtest_efi.lds boot/header.o boot/setup.o -b binary memtest_shared.bin -o memtest.efi
+       $(LD) --defsym=_bss_size=$(word 3,$(SIZES)) -T ldscripts/memtest_efi.lds boot/header.o boot/setup.o -Bdynamic memtest_shared -o memtest.efi
 
 memtest.mbr: memtest_shared.bin boot/mbr.o ldscripts/memtest_mbr.lds
        $(LD) -T ldscripts/memtest_mbr.lds boot/mbr.o -b binary memtest_shared.bin -o memtest.mbr
diff --git a/build64/ldscripts/memtest_efi.lds b/build64/ldscripts/memtest_efi.lds
index 14d1a49..18af6a0 100644
--- a/build64/ldscripts/memtest_efi.lds
+++ b/build64/ldscripts/memtest_efi.lds
@@ -13,11 +13,24 @@ SECTIONS {
        . = ALIGN(512);
        .text : {
                _file_text_start = . ;
-               *(.data)
+               *(.text)
                _real_text_end = . ;
                . = ALIGN(512);
                _file_text_end = . ;
        }
+       .data : {
+               _file_data_start = . ;
+               *(.rodata)
+               *(.rela.*)
+               . = ALIGN(4);
+               *(.data)
+               *(.got)
+               . = ALIGN(4);
+               *(.bss)
+               _real_data_end = . ;
+               . = ALIGN(512);
+               _file_data_end = . ;
+       }
        .reloc : {
                _file_reloc_start = . ;
                *(.reloc)
@@ -35,24 +48,28 @@ SECTIONS {
        /DISCARD/ : { *(*) }
 
        _real_text_size  = _real_text_end  - _file_text_start;
+       _real_data_size  = _real_data_end  - _file_data_start;
        _real_reloc_size = _real_reloc_end - _file_reloc_start;
        _real_sbat_size  = _real_sbat_end  - _file_sbat_start;
 
        _file_head_size  = _file_text_start;
        _file_text_size  = _file_text_end  - _file_text_start;
+       _file_data_size  = _file_data_end  - _file_data_start;
        _file_reloc_size = _file_reloc_end - _file_reloc_start;
        _file_sbat_size  = _file_sbat_end  - _file_sbat_start;
 
-       _sys_size  = (_real_text_size + 15) >> 4;
-       _init_size = _real_text_size + _bss_size;
+       _sys_size  = (_real_text_size + _real_data_size + 15) >> 4;
+       _init_size = _real_text_size + _real_data_size + _bss_size;
 
        _virt_head_size  = ((_file_head_size  + 4095) >> 12) << 12;
-       _virt_text_size  = ((_init_size       + 4095) >> 12) << 12;
+       _virt_text_size  = ((_file_text_size  + 4095) >> 12) << 12;
+       _virt_data_size  = ((_file_data_size  + 4095) >> 12) << 12;
        _virt_reloc_size = ((_file_reloc_size + 4095) >> 12) << 12;
        _virt_sbat_size  = ((_file_sbat_size  + 4095) >> 12) << 12;
 
        _virt_text_start  = _virt_head_size;
-       _virt_reloc_start = _virt_text_start  + _virt_text_size;
+       _virt_data_start  = _virt_text_start  + _virt_text_size;
+       _virt_reloc_start = _virt_data_start  + _virt_data_size;
        _virt_sbat_start  = _virt_reloc_start + _virt_reloc_size;
 
        _virt_img_size = _virt_sbat_start + _virt_sbat_size;
```
However, ld doesn't like it... right, memtest_shared is a shared object, not an object file, and `-Bdynamic` does not help:
```
ld --defsym=_bss_size=304032 -T ldscripts/memtest_efi.lds boot/header.o boot/setup.o -Bdynamic memtest_shared -o memtest.efi
ld: memtest_shared: error adding symbols: file in wrong format
make: *** [Makefile:177: memtest.efi] Error 1
```

**Comment by miczyg1** [comment]
> EDIT: and I guess that having a RWX text segment (additional IMAGE_SCN_MEM_WRITE flag, which could be an easy one-liner way out of this) is afoul some guidelines or requirements (e.g. related to Secure Boot, signed shims, etc.) ?

Having RWX on code section is definitely a bad idea. It would mean the code can modify itself. Not sure if paging would allow that too.

Regarding making the EFI executables from shared objects, found something like this:
https://wiki.osdev.org/GNU-EFI#Converting_Shared_Object_to_EFI_executable
Will try if it works. Not sure though if you would like gnu-efi in this project, but it may serve as some inspiration how to do linking right.

**Comment by martinwhitaker** [comment]
memtest86+ continually relocates itself whilst tests are running, so the code does modify itself.

**Comment by debrouxl** [comment]
Well, beyond the initial loading phase, memtest86+'s image does modify itself, when it copies and relocates itself back and forth between two memory locations, so that even the memory range where the loader initially put it can be tested as well.
For the same reason, memtest86+ exits the UEFI boot services early, when they're no longer needed: according to other issues in this tracker, on some motherboards, exiting the UEFI environment makes hundreds of additional MBs of memory available.

**Comment by miczyg1** [comment]
> memtest86+ continually relocates itself whilst tests are running, so the code does modify itself.

That said, it can relocate code and data (in separate sections) to a new location, set paging with proper attributes and protections and run there.  When relocating to a new location, initially whole code is treated as data, because it is simply copied. After copying it must be patched to execute with new base address. But nothing prevents from changing the page attributes to read-only after the code is relocated (assuming that data is a separate section and is copied accordingly to new base address).  Or am I missing something?

**Comment by martinwhitaker** [comment]
I was able to reproduce this issue in qemu by building a custom version of the EDK2 OVMF firmware with memory protection enabled. This has also shown up recently on real hardware platforms (issue #413). I have pushed the simple fix of marking  the .text section as both executable and writable to the main branch. That fixes the issue in both my qemu test and on the real hardware platforms, so I expect it to fix this issue too.
